rs/rust_canisters/xnet_test/src/main.rs

```diff
@@ -202,96 +181,103 @@ fn fanout() {
             continue;
         }
 
-        if canisters.contains(&self_id.get().as_slice().to_vec()) {
+        if canisters.contains(&self_id) {
             // Same subnet
             continue;
         }
 
+        let mut futures = vec![];
         for _ in 0..PER_SUBNET_RATE.with(|r| *r.borrow()) {
             let idx = RNG.with(|rng| rng.borrow_mut().gen_range(0..canisters.len()));
             let canister = canisters[idx].clone();
 
             let seq_no = STATE.with(|s| s.borrow_mut().next_out_seq_no(canister.clone()));
 
-            let msg = candid_encode_padded(&Request {
+            let payload_size = PAYLOAD_SIZE.with(|p| *p.borrow()) as usize;
+            let payload = Request {
                 seq_no,
                 time_nanos: time_nanos(),
-            });
+                padding: vec![0; payload_size.saturating_sub(16)],
+            };
 
-            let err_code = api::call_raw(
-                api::CanisterId::try_from(canister.clone()).unwrap(),
+            let res = call::<(Request,), (Reply,)>(
+                Principal::try_from(canister.clone()).unwrap(),
                 "handle_request",
-                &msg[..],
-                on_reply,
-                on_reject,
-                None,
-                std::ptr::null_mut(),
-                api::Funds::zero(),
+                (payload,),
             );
+            futures.push(res);
+        }
 
-            if err_code != 0 {
-                log(&format!(
-                    "{} call failed with {}",
-                    time_nanos() / 1_000_000,
-                    err_code
-                ));
-                METRICS.with(|m| m.borrow_mut().call_errors += 1);
-            } else {
-                METRICS.with(move |m| m.borrow_mut().requests_sent += 1);
+        let results = join_all(futures).await;
+
+        for res in results {
+            match res {
+                Ok((reply,)) => {
+                    let elapsed = Duration::from_nanos(time_nanos() - reply.time_nanos);
+                    METRICS.with(move |m| m.borrow_mut().requests_sent += 1);
+                    METRICS.with(|m| m.borrow_mut().latency_distribution.observe(elapsed));
+                }
+                Err((err_code, err_message)) => {
+                    // Catch whether the call failed due to a synchronous or
+                    // asynchronous error. Based on the current implementation of
+                    // the Rust CDK, a synchronous error will contain a specific
+                    // error message.
+                    if err_message.contains("Couldn't send message") {
+                        log(&format!(
+                            "{} call failed with {:?}",
+                            time_nanos() / 1_000_000,
+                            err_code
```

Comment:
FYI (no need to do anything about it): `err_code` used to be an `i32`. So while the log message was less informative (`call failed with 2` as opposed to `call failed with SysTransient`), you could also fit more of them in the limited size rotating log. I don't think it makes any difference in the big picture, though.

Reply:
I can get the number back, I just need to explicitly convert from `RejectCode` to the relevant raw number.

Reply:
Actually scratch that, I can't because we explicitly don't have that conversion. And I just remembered we don't want to have it because it can lead to other issues wrt encoding/decoding from the protobuf equivalent structs.